pkg/controller/client_builder_dynamic.go

```diff
@@ -0,0 +1,287 @@
+		clock:                clock.RealClock{},
+	}
+	return builder
+}
+
+func (t *DynamicControllerClientBuilder) Config(saName string) (*restclient.Config, error) {
+	_, err := t.getOrCreateServiceAccount(saName)
+	if err != nil {
+		return nil, err
+	}
+
+	configCopy := constructClient(t.Namespace, saName, t.ClientConfig)
+
+	ts, ok := t.get(saName)
+	if ok {
+		configCopy.TokenSource = ts
+	} else {
+		configCopy.TokenSource = &tokenSourceImpl{
+			clientConfig:       t.ClientConfig,
+			namespace:          t.Namespace,
+			serviceAccountName: saName,
+			coreClient:         t.CoreClient,
+			authClient:         t.AuthenticationClient,
+			clock:              clock.RealClock{},
+		}
+		t.set(saName, configCopy.TokenSource)
+	}
+
+	return &configCopy, nil
+}
+
+func (t *DynamicControllerClientBuilder) ConfigOrDie(name string) *restclient.Config {
+	clientConfig, err := t.Config(name)
+	if err != nil {
+		klog.Fatal(err)
+	}
+	return clientConfig
+}
+
+func (t *DynamicControllerClientBuilder) Client(name string) (clientset.Interface, error) {
+	clientConfig, err := t.Config(name)
+	if err != nil {
+		return nil, err
+	}
+	return clientset.NewForConfig(clientConfig)
+}
+
+func (t *DynamicControllerClientBuilder) ClientOrDie(name string) clientset.Interface {
+	client, err := t.Client(name)
+	if err != nil {
+		klog.Fatal(err)
+	}
+	return client
+}
+
+func (t *DynamicControllerClientBuilder) getOrCreateServiceAccount(name string) (*v1.ServiceAccount, error) {
+	sa, err := t.CoreClient.ServiceAccounts(t.Namespace).Get(name, metav1.GetOptions{})
+	if err == nil {
+		return sa, nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	// Create the namespace if we can't verify it exists.
+	// Tolerate errors, since we don't know whether this component has namespace creation permissions.
+	if _, err := t.CoreClient.Namespaces().Get(t.Namespace, metav1.GetOptions{}); err != nil {
+		t.CoreClient.Namespaces().Create(&v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: t.Namespace}})
+	}
+
+	// Create the service account
+	sa, err = t.CoreClient.ServiceAccounts(t.Namespace).Create(&v1.ServiceAccount{ObjectMeta: metav1.ObjectMeta{Namespace: t.Namespace, Name: name}})
+	if apierrors.IsAlreadyExists(err) {
+		// If we're racing to init and someone else already created it, re-fetch
+		return t.CoreClient.ServiceAccounts(t.Namespace).Get(name, metav1.GetOptions{})
+	}
+	return sa, err
+}
+
+func (t *DynamicControllerClientBuilder) get(key string) (oauth2.TokenSource, bool) {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+	ctr, ok := t.tokenSourceMap[key]
+	return ctr, ok
+}
+
+func (t *DynamicControllerClientBuilder) set(key string, tr oauth2.TokenSource) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	t.tokenSourceMap[key] = tr
+}
+
+type tokenSourceImpl struct {
+	clientConfig       *restclient.Config
+	namespace          string
+	serviceAccountName string
+	cache              *v1authenticationapi.TokenRequest
```

Comment:
Export caching token source and don't reimplement it here. Every time Token() gets called, just do a token review.

https://github.com/kubernetes/kubernetes/blob/989182435220f8c44c1cea67f44a795a5658ba66/staging/src/k8s.io/client-go/transport/token_source.go#L98